sdk/core/Azure.Core.Experimental/tests/public/JsonDataPublicTests.cs

```diff
@@ -0,0 +1,400 @@
+        public void DynamicCanConvertToIEnumerableDynamic()
+        {
+            dynamic jsonData = new BinaryData("[1, null, \"s\"]").ToDynamic();
+            int i = 0;
+            foreach (var dynamicItem in jsonData)
+            {
+                switch (i)
+                {
+                    case 0:
+                        Assert.AreEqual(1, (int)dynamicItem);
+                        break;
+                    case 1:
+                        Assert.AreEqual(null, (string)dynamicItem);
+                        break;
+                    case 2:
+                        Assert.AreEqual("s", (string)dynamicItem);
+                        break;
+                    default:
+                        Assert.Fail();
+                        break;
+                }
+
+                i++;
+            }
+            Assert.AreEqual(3, i);
+        }
+
+        [Test]
+        public void DynamicCanConvertToIEnumerableInt()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            int i = 0;
+            foreach (int dynamicItem in jsonData)
+            {
+                Assert.AreEqual(i, dynamicItem);
+
+                i++;
+            }
+            Assert.AreEqual(4, i);
+        }
+
+        [Test]
+        public void DynamicArrayHasLength()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            Assert.AreEqual(4, jsonData.Length);
+        }
+
+        [Test]
+        public void DynamicArrayFor()
+        {
+            dynamic jsonData = new BinaryData("[0, 1, 2, 3]").ToDynamic();
+            for (int i = 0; i < jsonData.Length; i++)
+            {
+                Assert.AreEqual(i, (int)jsonData[i]);
+            }
+        }
+
+        [Test]
+        public void CanAccessProperties()
+        {
+            dynamic jsonData = new BinaryData("{ \"primitive\":\"Hello\", \"nested\": { \"nestedPrimitive\":true } }").ToDynamic();
+
+            Assert.AreEqual("Hello", (string)jsonData.primitive);
+            Assert.AreEqual(true, (bool)jsonData.nested.nestedPrimitive);
+        }
+
+        [Test]
+        public void CanTestPropertyForNull()
+        {
+            dynamic jsonData = new BinaryData("{ \"primitive\":\"Hello\", \"nested\": { \"nestedPrimitive\":true } }").ToDynamic();
+
+            Assert.IsNull(jsonData.OptionalInt);
+            Assert.IsNull(jsonData.OptionalString);
+            Assert.AreEqual("Hello", (string)jsonData.primitive);
+        }
+
+        [Test]
+        public void CanAddStringToList()
+        {
+            dynamic jsonData = new BinaryData(new { value = "foo" }).ToDynamic();
+
+            List<string> list = new();
+
+            // TODO: Should we add an implicit cast to string?
+            // Can we do it dynamically?
+            // Alternatively, we could provide a different type for string leaf
+            // nodes, like we do for numbers.
+            list.Add(jsonData.value);
```

Comment:
This fails with a:

`Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : The best overloaded method match for 'System.Collections.Generic.List<string>.Add(string)' has some invalid arguments`

because jsonData.value's underlying type is JsonData, and there is no implicit conversion to string.  The DynamicMetaObject's BindConvert method is never invoked here, so we do not have the opportunity there to do the conversion.  Are we ok requiring a cast to string to make this succeed?  The cons to requiring the cast are that it is inconsistent with the usage of Azure SDK model types, and the compiler doesn't enforce it so developers won't discover the need for the cast until an exception is thrown at runtime.